compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -760,6 +760,18 @@ convertExpr env0 e = do
             recordType <- convertType env recordType
             record <- convertExpr env record
             pure $ ERecProj (fromTCon recordType) (mkField $ fsToText name) record
+    -- NOTE(SF): We also need to inline `setField` in order to get the correct
+    -- evaluation order (record first, then fields in order).
+    go env (VarIn DA_Internal_Record "setField") (LType (isStrLitTy -> Just name) : LType record@(TypeCon recordTyCon _) : LType field : _dict : args)
+        | isSingleConType recordTyCon = do
+            record' <- convertType env record
+            field' <- convertType env field
+            withTmArg env (varV1, field') args $ \x1 args ->
+                withTmArg env (varV2, record') args $ \x2 args ->
+                    pure (ERecUpd (fromTCon record') (mkField $ fsToText name) x2 x1, args)
+        -- TODO: Also fix evaluation order for sum-of-record types.
```

Comment:
In your PR description you mention that this will be a major challenge. Could you expand on that? I certainly see how it can be annoying to implement but I don’t see what exactly makes it challenging.

Reply:
I slept on it and I think there's an easy way to do it actually -- rewrite it as an `ELet` expression to force the "record" argument to be evaluated before the field. Before I thought it would have to branch on each constructor.